### PR TITLE
chore: block selected actions until the sync completes [AR-2443]

### PR DESCRIPTION
**Pull request by saleniuk**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-2443" title="AR-2443" target="_blank">BugAR-2443</a>  User can send messages while "connecting" is shown. Messages are lost into nowhere
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

User can send messages while "connecting" is shown. Messages are lost into nowhere. Not only sending messages is affected by that. When the user performs any action while the “connecting” is shown, lot’s of things fail.

### Solutions

Implement an easy to reuse solution to block some actions until the sync is completed.
Current sync state is provided using `CompositionLocalProvider`, checking for the state is possible by simply using a parameter `blockUntilSynced` added to WireButtons, WireIconButtons and clickable modifier using Clickable.

#### How to Test

Try to send a message or log out when "connecting" is still visible.

### Notes (Optional)

if anyone has a suggestion what additional actions should be blocked during sync, please let me know or write it down in a comment.

### Attachments (Optional)

I intentionally delayed the sync process for the video, which is why it takes so long there.

https://user-images.githubusercontent.com/30429749/192249491-e5fbee9b-24b4-44b8-a0c6-4f504d89c52d.mp4

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [1](https://10.10.124.134/job/wire-android-reloaded/job/PR-1038/1/) **failed**.

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-1038/2/) **failed**.

**Comment by AndroidBob** [comment]
Build [3](https://10.10.124.134/job/wire-android-reloaded/job/PR-1038/3/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(4522709)(4522709)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(4522709)(4522709)-dev-release.apk)

**Comment by AndroidBob** [comment]
Build [4](https://10.10.124.134/job/wire-android-reloaded/job/PR-1038/4/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(4539816)(4539816)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(4539816)(4539816)-dev-release.apk)

**Comment by AndroidBob** [comment]
Build [5](https://10.10.124.134/job/wire-android-reloaded/job/PR-1038/5/) **failed**.

**Comment by AndroidBob** [comment]
Build [6](https://10.10.124.134/job/wire-android-reloaded/job/PR-1038/6/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(4557383)(4557383)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(4557383)(4557383)-dev-release.apk)

**Comment by AndroidBob** [comment]
Build [7](https://10.10.124.134/job/wire-android-reloaded/job/PR-1038/7/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(4590401)(4590401)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(4590401)(4590401)-dev-release.apk)

**Comment by ohassine** [comment]
Great work @saleniuk, but I think this will be handled when we implement https://wearezeta.atlassian.net/browse/AR-2561 ?

**Comment by saleniuk** [comment]
> Great work @saleniuk, but I think this will be handled when we implement https://wearezeta.atlassian.net/browse/AR-2561 ?

AR-2561 will solve the issue with sending messages, but there are other actions that also should be disabled until the app is synced. 🤔

**Comment by AndroidBob** [comment]
Build [8](https://10.10.124.134/job/wire-android-reloaded/job/PR-1038/8/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(4616874)(4616874)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(4616874)(4616874)-dev-release.apk)

File: app/src/main/kotlin/com/wire/android/ui/common/Extensions.kt [hunk 1]
```diff
@@ -68,10 +72,28 @@ fun Modifier.shimmerPlaceholder(
...
+        onLongClick = clickable.onLongClick?.let { onLongClick ->
+            remember(clickable.blockUntilSynced) {
+                {
+                    if (clickable.blockUntilSynced && !syncStateObserver.isSynced)
+                        Toast.makeText(context, context.getString(R.string.label_wait_until_synchronised), Toast.LENGTH_SHORT).show()
+                    else
+                        onLongClick()
+                }
+            }
+        }
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/ui/common/Extensions.kt]: should we extract the clickable remember to a val for coherence with other composables?

File: app/src/main/kotlin/com/wire/android/ui/common/button/WireButton.kt [hunk 2]
```diff
@@ -77,8 +82,17 @@ fun WireButton(
...
+    val context = LocalContext.current
     Button(
-        onClick = onClick,
+        onClick = remember(blockUntilSynced) {
+            {
+                if (blockUntilSynced && !syncStateObserver.isSynced)
+                    Toast.makeText(context, context.getString(R.string.label_wait_until_synchronised), Toast.LENGTH_SHORT).show()
+                else
+                    onClick()
+            }
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/ui/common/button/WireButton.kt]: Same than above

**Review by gongracr — APPROVED** [review]
Looking great to me, I just added a tiny non-blocking commet. Even if we decide to fix the load of messages disabling them until sync is complete, I think this would still be useful, as we might have other exposed buttons that should be throttled until sync is complete too.

Great job 🍒

File: app/src/main/kotlin/com/wire/android/ui/common/Extensions.kt (hunk 1 above)
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/common/Extensions.kt]: Sure, already done ✔️

File: app/src/main/kotlin/com/wire/android/ui/common/button/WireButton.kt (hunk 2 above)
Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/ui/common/button/WireButton.kt]: Sure, already done ✔️

**Comment by AndroidBob** [comment]
Build [10](https://10.10.124.134/job/wire-android-reloaded/job/PR-1038/10/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(4643904)(4643904)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(4643904)(4643904)-dev-release.apk)

**Pull request merged**